### ISSUE TITLE
ModSecurity: collections_remove_stale: Failed deleting collection

### ISSUE THREAD

**Issue by rcbarnett** [SpiderLabs/ModSecurity#576]
MODSEC-428: When server is under load of 1500 concurrent users Modsecurity starts to show the below error in apache's error log and the server CPU usage goes up to dangerous levels.

[Wed Oct 02 21:21:52 2013] [error] [client 109.127.81.10] ModSecurity: collections_remove_stale: Failed deleting collection (name "ip", key "109.127.81.138_9e4f93d096e2ca3744251c41dde47a1a7b26fa75"): Internal error [hostname "www.regayzanko.com"] [uri "/php5-fcgi/setting/student_collage_list.php"] [unique_id "UkxkMMaIMEoAADg9zw0AAAgr"]

I can see that this issue was patched see: https://www.modsecurity.org/tracker/browse/MODSEC-97  but i am using the latest modsecurity version "2.7.5" so i do not know why it is happening.

**Comment by antoxxa** [comment]
stil no answer ?

**Comment by lifeforms** [comment]
I don't have this problem myself anymore since we're not using brute force checking on that server.

However, I found that the server was using `/tmp` to store the data files. This directory happened to be huge (75000 other entries). My suspicion is that it might take too long.

Could other people experiencing this issue try the following:

- Check where your data files are stored (`SecDataDir`) and see if this might be slow due to an overfull directory? Maybe cleaning up the `tmp` directory helps.
- Try setting `SecDataDir` to its own dedicated directory.
- Try setting `SecDataDir` to a tmpfs or memory filesystem.

I wonder if this makes the issue go away for you.

**Comment by antoxxa** [comment]
finding in SecDataDir  ip.pag and clearing it out helps ... but not sure why apache cannot do it itslef ... despite the fact that it is creating that file

**Comment by pgib** [comment]
@antoxxa Thanks. Had the exact same problem with an ip.pag file that had grown to 115MB.

**Comment by nobleclem** [comment]
Do we know if this is being worked on?  Seems like a but like this thats been open for about 2 years would have been investigated by now.

**Comment by onewaypub** [comment]
Same issue here with version 2.8.0. But error message is diffrent

ModSecurity: collection_retrieve_ex: Failed deleting collection (name "ip", key "217.254.155.4_6246985e080008849486b7f63d4773c0b4b3f2dd") ...

Also tried to delete the ip.pag file in the SecDataDir. But it do not help

**Comment by mamykin** [comment]
Has anyone seen this issue in 2.9.0? I don't have the time to test it.

**Comment by celesteking** [comment]
"ModSecurity: collections_remove_stale: Failed deleting collection XXX"

2.8.0, apache 2.2 mpm worker fixed thread count, suphp.
SecTmpDir writable by apache user, separate dir, (always empty?).
SecDataDir writable by apache user, separate dir.

Suggestions on solving this bug are welcome. Modsec devs, you might want to run a special separate cleaning thread/process or create another solution maybe?

**Comment by zimmerle** [comment]
Hi @celesteking,

You can cleanup your sdbm using a separated process: https://github.com/SpiderLabs/modsec-sdbm-util

Still, you may want to reduce the timeout and the amount of data to save in the collections. If you try to save a lot of data for a extended period of time, it eventually fails. There are some studies to use memcache to centralize the storage here:

https://github.com/SpiderLabs/ModSecurity/tree/memcache_collections

**Comment by celesteking** [comment]
That might work, thanks, we're gonna incorporate that util to run periodically.
SecCollectionTimeout has been set to 600 from day 1 here. Still, ip.pag file was growing up to 1gb.

**Comment by celesteking** [comment]
Speaking of, when modsec is in detectiononly mode, will the stale/expired entries be removed?

**Comment by zcwang3** [comment]
We are also having this issue with mod_security 2.9.0, any update/plan on this bug?
What's the real root cause for it?

**Comment by pgib** [comment]
I've had to resort to deleting the file in a cronjob nightly. O_o

**Comment by zcwang3** [comment]
if a IP address is detected to exceed the maximum number of IP request per Sec, request from this IP address will be blocked by Mod_security. 
How long does this IP address request continues to be blocked? 
Delete the file by cron job will impact the blocking behavior and then impact the preventing Dos attack effect?

refer to [Mod_security Notes](http://linux.icydog.net/apache/modsecurity-notes.php), collection's default expired time is one hour.
The documentation is very unclear about expiration of persistant collections. The documentation for expirevar seems to imply that collections by default do not expire. However, this is not the case. The **default expiration time is given later in the documentation as 3600 seconds (1 hour)**, but my first attempts at increasing the expiration time (useful for spam blacklists) failed mysteriously.

**Comment by dhaupin** [comment]
@zcwang3 Did you ever figure out what causes this issue? Just noticed that its happening on one of our srvs. Modsec 2.9.0

**Comment by zcwang3** [comment]
@dhaupin one of my colleague find that "collections_remove_stale: Failed deleting collection" happens when mod_security use **apr_sdbm_delete**() method to delete expired keys. Mod_security try to delete an sdbm record by key with Apache Portable Runtime Utility Library, but we do not know what causing this failure.

Issue existing in the storage of mod_security, alternative solution is change to use other storage like Redis or Memcached. Unfortunately, refer to [Implement Redis support as Collection backend on libmodsecurity](https://github.com/SpiderLabs/ModSecurity/issues/1139), it is not implemented.

Some one suggest using cron job to delete pag file every day and then restart apache, but it does not make sense to me. 

Since this issue exists 3 years without solution, and we find other similar issues, like: [IP persistence storage seems to not clean up ](https://github.com/SpiderLabs/ModSecurity/issues/574). We are **struggling** if it is reasonable to keep using mod_security in our production.

Any other suggestion to fix this issue is appreciated!

**related source code is as following**:
                if (expiry_time <= now) {
                    rc = **apr_sdbm_delete**(dbm, key);
                    if (rc != APR_SUCCESS) {
                            msr_log(msr, 1, "collections_remove_stale: Failed deleting collection (name \"%s\", "
                            "key \"%s\"): %s", log_escape(msr->mp, col_name),
                            log_escape_ex(msr->mp, key.dptr, key.dsize - 1), get_apr_error(msr->mp, rc));
                            msr->msc_sdbm_delete_error = 1;
                            goto error;
                    }

**Comment by zimmerle** [comment]
Hi @zcwang3,

We have already implemented support for LMDB [#1141] and InMemory back-end inside ModSecurity version 3. The support for Redis [#1139] and Memcache [#1140] are on our backlog, to be implemented soon.

I think the root cause of this problem is that we are abusing the usage of the **apr_sdbm** implementation. IHMO was not designed to be used so intensively. That is why we advice to reduce the expiration time to a lower value, therefore reducing the amount of register in a given collection. Anything that you can change in the rules in that sense will reduce the size of the collection.

The persistent collection keep a in-memory cache and from time to time it sync the memory content with the sdbm storage. Notice that each process from your webserver [worker, if you prefer] has a different in-memory collection that is synced with the sdbm from time to time. The error message that you saw, most likely is not really related to the size (in disk) of your collection. It is just saying that it is trying to delete a collection that was already deleted, probably by other process [that is one possibility].

The part of the code that calls the memory cleanup is available here:
https://github.com/SpiderLabs/ModSecurity/blob/master/apache2/modsecurity.c#L240-L253

There was a initial/experimental support for memcache built on top of 2.x, which is available here: 
https://github.com/SpiderLabs/ModSecurity/tree/v2/dev/memcache_collections

**Comment by marcstern** [comment]
Felipe,
From your explanation, it looks obvious that this warning should be disabled by default in 2.x.
The added value of this warning is nul.

**Comment by zimmerle** [comment]
Hi @marcstern, 

I don't think so, the warning is just a warning. It does not means that there is an error. In order circumstances the warning may be more meaningful.

**Comment by mturk** [comment]
Not sure what kind of mpm is used in case of corruption, but in case of threaded mpm's the file locking does not work for threads belonging to the same process. In such case threads can be doing different operations on the sdbm although those operations presume they are protected by lock. I have a patch (#1224) that solves persistent storage data corruption issues, so it could probably fix this one as well in case the issue is observed with threaded (worker, event) mpm's.

**Comment by nrob81** [comment]
Hello,

I had the same problem (debian 8.5, modsecurity 2.8.0).
I am disabled SecResponseBodyAccess and the warnings disappeared.

**Comment by bostrt** [comment]
This issue is resolved by @mturk's #1224 pull request. The issue is that multiple threads are trying to delete the same collection. I dropped in some debug logging to prove this.

I added the following (line offsets may be a bit off from master):
```diff
@@ -676,6 +699,11 @@ int collections_remove_stale(modsec_rec *msr, const char *col_name) {
                 }
 
                 if (expiry_time <= now) {
+                    if (msr->txcfg->debuglog_level >= 9) {
+                        msr_log(msr, 9, "collections_remove_stale: Attempting to delete collection (name \"%s\", "
+                            "key \"%s\")", log_escape(msr->mp, col_name),
+                            log_escape_ex(msr->mp, key.dptr, key.dsize - 1));
+                    }
                     rc = apr_sdbm_delete(dbm, key);
                     if (rc != APR_SUCCESS) {
                         msr_log(msr, 1, "collections_remove_stale: Failed deleting collection (name \"%s\", "
@@ -698,13 +726,14 @@ int collections_remove_stale(modsec_rec *msr, const char *col_name) {
```

I ran a simple test using serial requests and saw this logging output (scroll horizontally to see session ID):

```
[14/Feb/2017:22:25:28 --0500] [localhost/sid#5633ee000420][rid#5633ee144150][/index.html][9] collections_remove_stale: Attempting to delete collection (name "SESSION", key "session9271")
[14/Feb/2017:22:25:28 --0500] [localhost/sid#5633ee000420][rid#5633ee142140][/index.html][9] collections_remove_stale: Attempting to delete collection (name "SESSION", key "session9271")
[14/Feb/2017:22:25:28 --0500] [localhost/sid#5633ee000420][rid#5633ee142140][/index.html][9] collections_remove_stale: Attempting to delete collection (name "SESSION", key "session9271")
[14/Feb/2017:22:25:28 --0500] [localhost/sid#5633ee000420][rid#5633ee144150][/index.html][1] collections_remove_stale: Failed deleting collection (name "SESSION", key "session9271"): Internal error (specific information not available)
[14/Feb/2017:22:25:28 --0500] [localhost/sid#5633ee000420][rid#5633ee142140][/index.html][1] collections_remove_stale: Failed deleting collection (name "SESSION", key "session9271"): Internal error (specific information not available)
```

As you can see, the same `session9271` is attempting to be delete by 3 different threads. The first thread actually deletes it from storage and the subsequent two threads receive the "Failed deleting collection" message.

**Comment by alexgoldsilver** [comment]
The fix in PR 1224 has been outstanding for a while with no comments. These entries make parsing our audit logs for real threats very very difficult without heavily modifying our tools. Is there at least a way to suppress the warnings for this in the audit log using a configuration directive until some fix is applied to the threading issues?

**Comment by zimmerle** [comment]
@alexgoldsilver recently #1380 was merged into our mainline.

**Comment by zimmerle** [comment]
You may want to have a look at: #1224 by enabling the global mutex for collections, this problem should be resolved.
